containers/src/Data/Sequence/Internal.hs

```diff
@@ -1461,97 +1461,104 @@ before3FT firstx xs lastx =
     ffx = Elem firstx
     flx = Elem lastx
 
+-- Invariants for raptyMiddle:
+--
+-- 1. midxs is constant: the middle bit in the original sequence (xs = (first <: Seq midxs :> last))
+-- 2. lenys is constant: the length of ys
+-- 3. firstf and pr repeat the same element: the first one in the original sequence xs
+-- 4. lastf  and sf repeat the same element: the last  one in the original sequence xs
+-- 5. sizec = size firstf = size lastf
+-- 6. lenys = deep_count * sizec + size pr + size pf
+-- 7. let (lft, fill, rght) = fill23 x, for any x:
+--      7a. All three sequences repeat the element x
+--      7b. size fill = sizec
+--      7c. size lft  = size sf
+--      7d. size rght = size pr
+-- 8. size result = deep_count * sizec + lenys * (size midxs + 1)
 raptyMiddle
-  :: Int
-  -> c
-  -> c
-  -> (a -> (Node c, Node c, Node c))
-  -> FingerTree (Elem a)
-  -> RCountMid c
-  -> FingerTree (Node c)
+  :: FingerTree (Elem a)  -- midxs
+  -> Int                  -- lenys
+  -> Int                  -- sizec
+  -> Node c               -- firstf
+  -> Node c               -- lastf
+  -> (a -> (Node c, Node c, Node c))  -- fill23
+  -> RCountMid c          -- (RCountMid pr deep_count sf)
+  -> FingerTree (Node c)  -- result
 
 -- At the bottom
 
-raptyMiddle !_sizec
-           _firstf
-           _lastf
-           fill23
-           fs
-           (RCountMid s pr 0 sf)
-     = deep
+raptyMiddle midxs lenys
+            !_sizec
+            _firstf
+            _lastf
+            fill23
+            (RCountMid pr 0 sf)
+     = Deep (lenys * (size midxs + 1))
             (One pr)
-            (mapMulFT s swizzle fs)
+            (mapMulFT lenys swizzle midxs)
             (One sf)
    where
      -- swizzle ::  Elem a -> Node (Node c)
      swizzle (Elem a) = case fill23 a of
         (lft, _fill, rght) -> Node2 (size pr + size sf) lft rght
 
-raptyMiddle sizec
-           firstf
-           lastf
-           fill23
-           fs
-           (RCountMid s pr 1 sf)
-     = deep
-            (Two pr (Node3 (3 * sizec) firstf firstf firstf))
-            (mapMulFT s swizzle fs)
-            (Two (Node3 (3 * sizec) lastf lastf lastf) sf)
+raptyMiddle midxs lenys
+            !sizec
+            firstf
+            lastf
```

Comment:
We should probably rename `firstf -> firstx` and `lastf -> lastx`. Could you amend that to this commit?